商品出品機能2

naoya1215 opened the pull request:
# What
商品出品機能の作成

#Why
商品機能を実装するため

①ログイン状態の場合は、商品出品ページへ遷移できる動画
https://gyazo.com/57dc084127ce21c45843fa69f38e2a86

②価格が入力されると同時に、販売手数料と販売利益が表示される動画
https://gyazo.com/f934e83e637fc098a7dc1b19772c2cee

③ 必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存される動画
https://gyazo.com/b478758938c2c765439089adf1c09418

④入力に問題がある状態で「出品する」ボタンが押された場合、情報は保存されず、出品ページに戻りエラーメッセージが表示される動画
https://gyazo.com/259cabe08460a215a6f4e98de90d4648

⑤ ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/114741c9a248fc7888875a65314ab15f

⑥テスト結果の画像
https://gyazo.com/6960cfa927d245da914e687d6b87949b

naoya1215 committed:
itemテーブルの作成

naoya1215 committed:
出品画面等の実装

naoya1215 committed:
Active Storageの導入

naoya1215 committed:
ActiveHashの実装

naoya1215 committed:
ログインページへの自動移行

ログインしていないユーザーの場合は、ログイン画面に促す

naoya1215 committed:
データベースへの保存

必要な情報を適切に入力して「出品する」ボタンを押すと、商品情報がデータベースに保存されること。

naoya1215 committed:
TOP画面の表示

naoya1215 committed:
バリデーションの追加

naoya1215 committed:
エラーメッセージの表示

naoya1215 committed:
正常系テストコード

正常系のテストコードの実装

naoya1215 committed:
異常系テストコード

商品出品機能の異常系テストコード実装

Akinori0123 commented on app/controllers/application_controller.rb:
```diff
@@ -1,6 +1,6 @@
 class ApplicationController < ActionController::Base
-  # before_action :authenticate_user! #ログインしていないユーザーをログインページの画面に促すことができる
-  before_action :basic_auth # Basic認証
+  before_action :authenticate_user! #ログインしていないユーザーをログインページの画面に促すことができる
```
```
before_action :authenticate_user! #ログインしていないユーザーをログインページの画面に促すことができる
```
上記は削除しましょう。

理由
・application_controller.rbに記載をすると今後作成する全てのコントローラのアクションに作用してしまうため
・items_controller.rbに既にauthenticate_user!の記載があるため

Akinori0123 commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,27 @@
+  before_action :authenticate_user!, except: [:index, :show]
+
+  def index #トップ画面の表示
+    @items = Item.all
```
```
@items = Item.all
```
上記をコメントアウトか削除しましょう。
次の一覧機能の内容であり、現時点で実装をしてしまうと一覧機能ブランチで差分を作れなくなってしまうためです。

Akinori0123 commented on app/views/items/index.html.erb:
```diff
@@ -143,17 +144,18 @@
           </div>
         </div>
         <% end %>
+      <% end %>
```
**index.html.erbの変更点**
出品するボタンのパス以外の変更点を全て配布時の状態に戻しましょう。
（該当箇所：133、136、147、150、158行目）
これらも次の一覧機能の内容であり、現時点で実装をしてしまうと一覧機能ブランチで差分を作れなくなってしまうためです。

*素材フォルダから再度index.html.erbを配置し直して出品するボタンのパスのみ再度実装する形が確実かと考えられます。

Akinori0123 commented on app/views/items/new.html.erb:
```diff
@@ -0,0 +1,159 @@
+
+    <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
+    <%= render 'shared/error_messages', model: f.object %>
+    <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
```
<%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
<%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
上記実装案内コメントも内容を実装済みのため削除しましょう。

Akinori0123 commented on spec/factories/items.rb:
```diff
@@ -0,0 +1,19 @@
+FactoryBot.define do
+  factory :item do
+    image {Faker::Lorem.sentence}
```
```
image {Faker::Lorem.sentence}
```
上記は削除しましょう。

理由：
・画像はactivestorageを使用しており文字列を代入することができないため
・16行目で画像を用意済みのため

Akinori0123 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,98 @@
+        @item.shipping_id = ''
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Shipping can't be blank")
+      end
```
**商品の状態, 配送料の負担, 発送元の地域, 発送までの日数, カテゴリーのカラムのテスト**
要件を満たしていることを確認するために、上記5カラムそれぞれに対して**未選択（'--'に該当するid）だと登録できないこと**もテストしましょう。

Akinori0123 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,98 @@
+        @item.price = 10000000
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Price must be less than or equal to 9999999")
+      end
```
**価格のテスト**
下限上限のテストの実装とても良いですね。
要件を満たしていることを確認するために、**価格に半角英数字以外が含まれていると登録できないこと**もテストしましょう。

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
コミット内容と添付いただいたgyazoの方確認いたしました。
コメントいたしましたのでご確認の程よろしくお願いいたします。

naoya1215 committed:
ログインページ画面に促すコードを削除

・application_controller.rbに記述されていると今後作成する全てのコントローラのアクションに作用してしまうため。
・items_controller.rbに既にauthenticate_user!の記載があるため

naoya1215 committed:
コメントアウト(@items = Item.all)

次の一覧機能の内容であり、現時点で実装をしてしまうと一覧機能ブランチで差分を作れなくなってしまうためです。

naoya1215 committed:
出品するボタンのパス以外の変更点を全て戻す

これらも次の一覧機能の内容であり、現時点で実装をしてしまうと一覧機能ブランチで差分を作れなくなってしまうため

naoya1215 committed:
コメント修正

コメントを削除し、わかりやすいコメントを追加

naoya1215 committed:
不要なfactory_botの削除

・画像はactivestorageを使用しており文字列を代入することができないため
・画像を用意済みのため

naoya1215 committed:
異常系のテストコード修正

商品の状態, 配送料の負担, 発送元の地域, 発送までの日数, カテゴリーのカラムのテスト
要件を満たしていることを確認するために、上記5カラムそれぞれに対して未選択（'--'に該当するid）だと登録できないこともテスト

naoya1215 committed:
価格のテスト追加

価格に半角英数字以外が含まれていると登録できないこと

naoya1215 committed:
正常系のテストコード修正(商品名、商品説明)

何度かテストを実施するとエラーが発生してしまうため修正を行う。

naoya1215 commented:
早速のご確認有難うございます。
コメントいただきました部分を修正しましたのでお手数ではございますが再度ご確認をお願いします。
以上となりますが、よろしくお願いします。

テスト結果のみ再度お送りします。
https://gyazo.com/f9db21f566fd10c1ca405c90104f4300

Akinori0123 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,105 @@
+        expect(@item.errors.full_messages).to include("Explanation can't be blank")
+      end     
+      it "category_id(カテゴリ)が未選択だと登録できない" do
+        @item.category_id = '---'
```
```
@item.category_id = '---'
```
category_idにはid番号を格納するため、異常値には'---'ではなくモデルで定めた'---'に該当するid番号を代入しましょう。
（その他の「未選択だと登録できない」テストにつきましても同様に修正しましょう）

Akinori0123 commented on app/javascript/price.js:
```diff
@@ -0,0 +1,14 @@
+const priceInput = document.getElementById("item-price");
+  priceInput.addEventListener("input", () => {
+    const inputValue = priceInput.value;
+    console.log(inputValue);
```
```
console.log(inputValue);
```
デバッグ用のconsole.logは今後使用しなければ削除しましょう。

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
前回の修正確認いたしました。追加でコメントいたしましたのでご確認いただき、修正後改めてレビュー依頼をいただけますと幸いです。
よろしくお願いいたします。

また、修正後テスト結果のみ再度gyazoでご共有いただけますと幸いです。併せてよろしくお願いいたします。

naoya1215 committed:
異常値の修正('---'→'1')

category_idにはid番号を格納するため、異常値には'---'ではなくモデルで定めた'---'に該当するid番号を代入しましょう。
（その他の「未選択だと登録できない」テストにつきましても同様に修正しましょう）

naoya1215 committed:
不要部分を削除

console.logの削除

naoya1215 commented:
有難うございます、修正ましたので確認をお願いします。

テスト結果
https://gyazo.com/f09bf8afa28309a8a867c88289a69979

Akinori0123 commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,105 @@
+        expect(@item.errors.full_messages).to include("Explanation can't be blank")
+      end     
+      it "category_id(カテゴリ)が未選択だと登録できない" do
+        @item.category_id = '1'
```
【修正任意（修正後のレビュー依頼は不要となります）】
```
@item.category_id = '1'
```
idが1番の時に登録できないテストの実装とても良いですね。
格納されるのがid番号で数値になりますので、より厳密には以下の方が適切となります。
```
@item.category_id = 1
```
結果としては変わらないかと考えられますので、マージ前にご確認だけお願いできればと思います。

Akinori0123 reviewed (commented):
レビュー依頼いただきありがとうございます！
前回の修正確認いたしました。LGTMです。

軽微な内容でコメントいたしましたのでマージ前にご確認いただけますと幸いです。
（修正後のレビュー依頼は不要となります）
実装お疲れ様でした。引き続き学習頑張っていきましょう！

naoya1215 committed:
'1'→1に修正

メンターからのLGTM前のアドバイス
